### PR TITLE
Nested settings

### LINKED ISSUE #33: Add support for nested settings
It would help the readability if `django-appsettings` supported nested settings out of the box. Consider, that I have complicated app with lots of settings and I want to split these settings into groups like this:

```python
MYAPP_API = {
    'SERVER': 'localhost',
    'PORT': 80,
    'SOME_LIST_SETTING': [...],
    ...
}
```

I'd like to define that I have nested setting `MYAPP_API` which consists of several other settings (in this case `StringSetting server`, `IntegerSetting port` and `ListSetting some_list_setting`. Each of these subsettings should be validated separately. Not all of them has to be present, depending on `required` param for each of these subsettings. My settings could then look like this:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        settings={
            'server': appsettings.StringSetting(required=True),
            'port': appsettings.IntegerSetting(default=80),
            'username': appsettings.StringSetting(),
            ...
        }
    )

    class Meta:
        setting_prefix = 'myapp_'
```

Checker would then validate each inner setting and the nested setting itself (it would be required if and only if any of it's inner settings is required). I don't know whether there are any relevant arguments that you may want to pass to the `NestedSetting`. If there aren't, we could even use pretty kwargs syntax:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
            server=appsettings.StringSetting(required=True),
            port=appsettings.IntegerSetting(default=80),
            username=appsettings.StringSetting(),
            ...
        }
    )

    class Meta:
        setting_prefix = 'myapp_'
```

### PR DESCRIPTION
**Pull request by stinovlas**
Closes #33.

I implemented the `NestedSetting` class as described in #33. I had to do some minor changes to `Setting` class, but overall, the change is pretty well contained. I also wrote tests for the updated and new code.

I'm not sure about how to describe the new class in the docs. I updated the main example, but maybe `NestedSetting` deserves a chapter on its own. What do you think? Otherwise, I followed contributing guidelines.

Also, `NestedSetting` can be nested itself (with arbitrary depth), but currently, there is no test for this behaviour and it's not well documented. Do you think that we should fix this?

Btw. the build is failing in several stages, but it's also failing on master and I don't think that any of these errors are caused by the changes I made. The tests themselves are passing.

### CODE REVIEW

**Comment by pawamoy** [comment]
Wow it was quick :smiley: ! About the failing stages, I'm gonna fix them and then you'll be able to rebase your branch. Sorry for that.

File: src/appsettings/settings.py [hunk 1]
```diff
@@ -872,3 +881,74 @@ def transform(self, path):
         for obj in objects:
             current_object = getattr(current_object, obj)
         return current_object
+
+
+# Nested settings -------------------------------------------------------------
+class NestedSetting(DictSetting):
+    """Nested setting."""
+
+    def __init__(self, settings, name='', default=lambda: dict(),
```
Comment (pawamoy) [inline src/appsettings/settings.py]: As told by Codacy, the `lambda: dict()` is not necessary here, you can directly use `dict` which is callable.

**Comment by stinovlas** [comment]
As for Codacy issue, I can replace `default=lambda: dict()` with `default=dict`, in our case, it gives the same result. But lamdas are used in lots of other `Setting` classes, so maybe it would be better to stick with them, for the sake of consistency. Or we can do the replacement in all classes. That way, Codacy won't complain and code will stay consistent.

File: src/appsettings/settings.py (hunk 1 above)
Comment (pawamoy) [reply src/appsettings/settings.py]: Ah! You're too quick for me! You fixed them anyway? Good :rofl:

Comment (stinovlas) [reply src/appsettings/settings.py]: Well, yes :sweat_smile:.

**Comment by stinovlas** [comment]
> About the docs, I think that indeed a more complete example of usage of NestedSetting would be good.
>
> I think your proposal in the issue conversation can be used (with maybe a bit of rewording) as it already contains all the explanation of the NestedSetting behavior.
>
> Would you mind adding such examples/explanation in the Using the settings in your code part of the documentation, as a special case or something like that?

Sure, I'll get to it.

> About the arbitrary depth nesting, I think it's okay if we don't mention it in the docs or add tests right now, unless you plan to use it immediately.

I don't plan to use it immediately, we can wait :slightly_smiling_face:.

**Comment by stinovlas** [comment]
One more thing: currently, the code allows settings to contain other items than those defined in appsettings class. 

```python
class MySetting(appsettings.AppSettings):
    nested = appsettings.NestedSetting(settings={})
```

```python
NESTED={'A': 0, 'B': 1}
```

So far, this configuration is valid and those additional items are ignored. However, maybe we should consider it to be invalid configuration and throw and exception. What do you think?

**Comment by pawamoy** [comment]
Hmmm... I think we don't want to be too smart. If users want to add extra stuff, let's let them. It's like adding an attribute to a class instance: `a = A(); a.unplanned_attribute = True`, it's not really good practice but it's not forbidden.

Maybe we can also copy others: what happens in Django if we add an extra key into, for example, one of the database in `DATABASES` setting?

**Comment by stinovlas** [comment]
> Maybe we can also copy others: what happens in Django if we add an extra key into, for example, one of the database in `DATABASES` setting?

It seems that nothing happens, not even a warning. Lets leave it at that.

**Comment by pawamoy** [comment]
The example is very clear, well done :slightly_smiling_face:. Thanks a lot for this great work @stinovlas, I think we can merge immediately once the tests are OK, unless you wanted to add something more!

**Comment by stinovlas** [comment]
Yes, you can merge it. Thank you :slightly_smiling_face:. Btw. don't forget to autosquash the commits, there is one fixup among them.

**Comment by stinovlas** [comment]
Hello again :-). Would you mind issuing new release to PyPI, which includes this PR?

**Comment by pawamoy** [comment]
Hey, sorry for the delay! Yes of course, gonna do this right now.

**Pull request merged**